### DISCUSSION TITLE
Is it possible to save audit events as part of Entity Framework transaction?

### DISCUSSION THREAD

**Discussion by voroninp** [thepirat000/Audit.NET#716] [Q&A]
My aggregate is publishing events.
I don't want to audit the entities (thus not using Audit EF package), but I want to save published events as audit events within the same transaction.

Is it possible?

**Comment by thepirat000** [comment]
I don't understand the issue, you have to provide more information, and a code sample explaining what are you trying to accomplish and what you've tried so far. 

What do you mean by "My aggregate is publishing events"?

**Reply by voroninp** [reply]
By aggregate I mean the implementation of DDD tactical pattern.
Aggregate is a unit of consistency and concurrency and is represented by one or several entities (in terms of Entity Framework).
State change happens only via method calls, and there's an internal collection which stores domain events. These events are created and added to the collection by the methods which change aggregate's state.

When `SaveChanges` is called interceptor scans the tracked aggregate and persists the events added to the collection.

In other words, I do not want to audit all EF entities, I just want to audit persisted events.

**Reply by thepirat000** [reply]
I understand the concept, but it's still unclear what you're trying to accomplish. Do you have a custom EF interceptor managing the internal collection?

The provided SaveChanges interceptor audits the entities from the EF change tracker. Is your internal collection being tracked by the EF change tracker?

Without code examples, it will be quite difficult to provide meaningful assistance.

**Reply by voroninp** [reply]
Here's the code for the aggregate:

```
[AggregateRoot]
[Entity]
public abstract class AggregateRootBase : IAggregateRoot
{
    private readonly List<object> _events = new();
    public IReadOnlyCollection<object> Events { get; }

    /// <summary>
    /// Add event to the queue of events for subsequent publishing on save or delete.
    /// </summary>
    /// <remarks>
    /// When and how events get published, is an implementation detail.
    /// </remarks>
    protected void EnqueueEventForPublishing(object @event)
    {
        _events.Add(@event);
    }

    protected AggregateRootBase()
    {
        Events = new ReadOnlyCollection<object>(_events);
    }

    public int Version { get; protected set; }

    private bool _aggregateHasChanged;
    bool IAggregateRoot.AggregateHasChanged => _aggregateHasChanged;

    protected void FlagAsChanged() => _aggregateHasChanged = true;

    public void AfterPersisted()
    {
        _aggregateHasChanged = false;
        _events.Clear();
        OnAfterPersisted();
    }

    public virtual void OnAfterPersisted() { }
}
```

Currently, the events are not persisted. On `SaveChanges` they just trigger handlers which work within the same transaction:
```
/// <summary>
/// Publishes event, that is initiates event processing by interested handlers and returns result of the handling:
/// Success or Failure.
/// </summary>
/// <remarks>
/// Publishing occurs within current transaction, if transaction fails, 
/// all changes introduced as a result of handling will be rolled back.<br/><br/>
/// 
/// Not to be confused with publishing messages for integration between services.
/// </remarks>
public interface IBeforeCommitDomainEventPublisher
{
    Task<Result> Publish(object domainEvent, CancellationToken cancellationToken = default);
}
```

Now I am thinking about building an audit log using your library and events (inversed EventSourcing in some sense :-) )
```
AuditScope.Log(@event)
```
Obviously, audit events should be persisted in the same transaction as changes to the aggregate are.

**Reply by thepirat000** [reply]
I believe I understand now, if my assumptions are correct, you should be able to do something like this:

```c#
await using (var transaction = await dbContext.Database.BeginTransactionAsync())
{
    await dbContext.SaveChangesAsync();

    await StoreCustomAuditEventsAsync(dbContext); 

    await transaction.CommitAsync();
}
```

The transaction will not be committed if either `SaveChanges` or `PublishEvents` fails.

**Reply by voroninp** [reply]
But will Audit.NET data provider enlist in this transaction?

**Reply by thepirat000** [reply]
Not automatically, but it depends on the data provider you plan to use and the location where the events are stored. Are the audit events going to be stored in the same database as your data?

**Reply by voroninp** [reply]
Yes, the same database. My initial assumption was that EF provider supports this, but then I saw the following comment:

> Store EntityFramework audit events in the same EF context. (This data provider **can only be used for Entity Framework audits**)

**Reply by thepirat000** [reply]
I'm creating a generic DbContext Data Provider to enable the storage of AuditEvents using an Entity Framework DbContext. This provider will support storing any type of AuditEvent and should align well with your use case. I'll let you know once it's released

**Comment by thepirat000** [comment]
The latest version 27.3.1 includes a [new DbContext DataProvider](https://github.com/thepirat000/Audit.NET/blob/master/src/Audit.EntityFramework/README.md#generic-dbcontext-data-provider) as a generic audit data provider that uses Entity Framework Core to write audit events. It allows mapping audit events to a specific entity in a DbContext.

I think this can be helpful for your use case. 

----

In the following example, the `MyService` class handles database operations and logs an audit event within a single transaction through its `ExecuteAsync` method.

The `MyAuditLogger` is responsible for generating custom logs. It leverages the injected `IAuditScopeFactory` to create and persist custom audit events. Its `LogAsync` method utilizes the new `DbContextDataProvider` to save the audit event, reusing the existing DbContext instance consequently ensuring they are part of the same transaction.

The `MyDbContext` serves as an example DbContext that manages data (`Orders`) while also storing audit events (`AuditLogs`).

```c#
using Audit.Core;
using Audit.EntityFramework.Providers;

using Microsoft.EntityFrameworkCore;
using Microsoft.Extensions.DependencyInjection;

// *************
// Startup Code:
// *************

// Register the services
var serviceProvider = new ServiceCollection()
    .AddScoped<IMyService, MyService>()
    .AddScoped<IMyAuditLogger, MyAuditLogger>()
    .AddSingleton<IAuditScopeFactory, AuditScopeFactory>()
    .AddDbContextFactory<MyDbContext>()
    .BuildServiceProvider();


// ***************************************
// Simulate a MyService.ExecuteAsync call:
// ***************************************
var service = serviceProvider.GetService<IMyService>();
await service.ExecuteAsync();


// *************
// Service Code:
// *************
public interface IMyService
{
    Task ExecuteAsync();
}

public class MyService : IMyService
{
    private readonly IDbContextFactory<MyDbContext> _contextFactory;
    private readonly IMyAuditLogger _auditLogger;

    public MyService(IDbContextFactory<MyDbContext> contextFactory, IMyAuditLogger auditLogger)
    {
        _contextFactory = contextFactory;
        _auditLogger = auditLogger;
    }

    public async Task ExecuteAsync()
    {
        // Create a new DbContext instance
        await using var dbContext = await _contextFactory.CreateDbContextAsync();

        // Start a new transaction
        await using var tran = await dbContext.Database.BeginTransactionAsync();

        // Add a new Order
        await dbContext.Orders.AddAsync(new Order()
        {
            OrderDate = DateTime.Now,
            OrderStatus = "C",
            OrderLines =
            [
                new OrderLine() { ProductCode = "A1", Quantity = 2 },
                new OrderLine() { ProductCode = "B2", Quantity = 3 }
            ]
        });

        // Save the database changes
        await dbContext.SaveChangesAsync();

        // Log the audit event within the same transaction
        await _auditLogger.LogAsync("LogTest", new { CustomField = "Test" }, dbContext);

        // Commit the transaction
        await tran.CommitAsync();

    }
}

public interface IMyAuditLogger
{
    Task LogAsync(string eventType, object customFields, MyDbContext dbContext);
}

public class MyAuditLogger : IMyAuditLogger
{
    private readonly IAuditScopeFactory _auditScopeFactory;

    public MyAuditLogger(IAuditScopeFactory auditScopeFactory)
    {
        _auditScopeFactory = auditScopeFactory;
    }

    // Your custom log method
    public async Task LogAsync(string eventType, object customFields, MyDbContext dbContext)
    {
        // Logic is similar to the static AuditScope.LogAsync(), but using the audit scope factory
        // and setting the data provider to save the audit logs using MyDbContext and AuditLog

        await _auditScopeFactory.CreateAsync(f => f
            .EventType(eventType)
            .ExtraFields(customFields)
            .IsCreateAndSave()
            .DataProvider(new DbContextDataProvider<MyDbContext, AuditLog>(c => c
                .DbContext(dbContext)
                .Mapper((auditEvent, auditLog) =>
                {
                    auditLog.AuditDate = DateTime.Now;
                    auditLog.Data = auditEvent.ToJson();
                }))));
    }
}

// Sample DbContext including Orders and AuditLogs
public class MyDbContext : DbContext
{
    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseSqlServer("data source=localhost;initial catalog=Audit_10;integrated security=true;Encrypt=False;");
    }

    public DbSet<Order> Orders { get; set; }

    public DbSet<AuditLog> AuditLogs { get; set; }
}

public class Order
{
    public int Id { get; set; }
    public DateTime OrderDate { get; set; }
    public string OrderStatus { get; set; }

    public List<OrderLine> OrderLines { get; set; }
}

public class OrderLine
{
    public int Id { get; set; }
    public int Quantity { set; get; }
    public string ProductCode { set; get; }
    public int OrderId { get; set; }
    public Order Order { get; set; }
}


public class AuditLog
{
    public int Id { get; set; }
    public DateTime AuditDate { get; set; }
    public string Data { get; set; }
}
```

Hope it helps

**Reply by voroninp** [reply]
Does it use context's `SaveChanges` or raw sql under the hood? Because if it uses `SaveChanges` then it risks persisting other tracked entities.

Is it possible to change the behavior to just attaching audit events to the change tracker, so they are saved on call to `SaveChanges`?

**Reply by thepirat000** [reply]
Yes, the `DbContextDataProvider` indeed calls `SaveChanges`. 

Additionally, the `SqlDataProvider` available in the `Audit.NET.SqlServer` package can be utilized. It also leverages a `DbContext` but operates with raw SQL.

Regarding your last question, I believe you can simply use Entity Framework for that purpose—you might not even need the Audit library, correct?

```c#
public async Task ExecuteAsync()
{
    await using var dbContext = await _contextFactory.CreateDbContextAsync();

    await dbContext.Orders.AddAsync(new Order()
    {
        OrderDate = DateTime.Now,
        OrderStatus = "C"
    });

    await dbContext.AuditLogs.AddAsync(new AuditLog()
    {
        AuditDate = DateTime.Now,
        Data = "event data..."
    });

    await dbContext.SaveChangesAsync();
}
```

**Reply by voroninp** [reply]
> Regarding your last question, I believe you can simply use Entity Framework for that purpose—you might not even need the Audit library, correct?

Yes and no. I like that Audit.Net produces audit events containing environment information. I'd like to use this machinery as it already exists in your library.
